### PR TITLE
Add full-speed USB OTG HS support

### LINKED ISSUE #8791: Request usb_dc_stm32 to support OTG_HS port on STM32F4 SoCs
The OTG_HS port on STM32F4 SoCs can support both full-speed USB transmission with its integrated transceiver at 12Mbps, and high-speed USB transmission at 480Mbps with the external USB phy.  So, the pins PB14 and PB15 from OTG_HS port can be directly used as a full-speed USB interface, providing same functionality as the OTG_FS port from PINs PA11 and PA12. To support OTG_HS as a full-speed USB port, the device driver usb_dc_stm32 needs to be changed to use the corresponding registers for OTG_HS port.

### PR DESCRIPTION
**Pull request by ydamigos**
Enable the driver to work in full-speed mode on OTG HS 
controller using its full-speed or high-speed internal PHY.

It allows to add usb support to boards like stm32f429i_disc1.

Tested on stm32f429 SoC by @jli157 and on STM32F723E-DISCO by @aurel32 .

Fixes #8791

Please note that only one interface should be enabled
at a time, OTG FS or OTG HS. The driver will raise an
error if both are enabled.

### CODE REVIEW

**Comment by ydamigos** [comment]
@jli157 and @aurel32 could you also review it once more?

File: drivers/usb/device/usb_dc_stm32.c [hunk 1]
```diff
@@ -173,9 +181,15 @@ static int usb_dc_stm32_clock_enable(void)
 	struct device *clk = device_get_binding(STM32_CLOCK_CONTROL_NAME);
 	struct stm32_pclken pclken = {
 
+#ifdef CONFIG_USB_HS_BASE_ADDRESS
+		.bus = STM32_CLOCK_BUS_AHB1,
+		.enr = LL_AHB1_GRP1_PERIPH_OTGHS
+#else /* CONFIG_USB_HS_BASE_ADDRESS */
+
 #ifdef USB
```
Comment (jli157) [inline drivers/usb/device/usb_dc_stm32.c]: In which case will the macro 'USB' be enabled?

Comment (ydamigos) [reply drivers/usb/device/usb_dc_stm32.c]: STM32Cube HAL defines USB on SoCs with USB device peripheral (not OTG FS or OTG HS) e.g. STM32103RB (see section 23 in RM0008).

**Comment by aurel32** [comment]
Thanks for making this PR, it looks all good to me, and works fine on my board. There is just a small conflict to solve now that usbotgfs support has been pushed.

**Comment by ydamigos** [comment]
Updated to resolve some conflicts.

**Review by aurel32 — APPROVED** [review]
Thanks for the rebase, it looks all good to me.

**Review by jli157 — APPROVED** [review]
Looks good to me! Thanks very much for making this PR!

**Comment by jli157** [comment]
Any chance to approve and merge this PR if no further issue is found?

**Review by finikorg — COMMENTED** [review]
Tested on stm32f723e_disco

I have added fix for FS mode.
For the HS mode I have made following change:
```
--- a/boards/arm/stm32f723e_disco/stm32f723e_disco.dts
+++ b/boards/arm/stm32f723e_disco/stm32f723e_disco.dts
@@ -61,6 +61,6 @@
        status = "ok";
 };
 
-&usbotg_fs {
+&usbotg_hs {
        status = "ok";
 };
```
And connected to other USB connector USB-HS. It works but apparently at FS. My Beagle 5000 sniffer shows <Full-speed> mode. And Linux also shows that "new full-speed device added".

**Comment by ydamigos** [comment]
> I have added fix for FS mode.

Thanks

> It works but apparently at FS. My Beagle 5000 sniffer shows mode. And Linux also shows that "new full-speed device added".

PR adds support for OTG HS controller for full speed mode. You could try the following change and test if it works in high speed:
```--- a/drivers/usb/device/usb_dc_stm32.c
+++ b/drivers/usb/device/usb_dc_stm32.c
@@ -281,7 +281,7 @@ static int usb_dc_stm32_init(void)
        usb_dc_stm32_state.pcd.Instance = USB_OTG_FS;
 #endif
        usb_dc_stm32_state.pcd.Init.dev_endpoints = CONFIG_USB_NUM_BIDIR_ENDPOINTS;
-       usb_dc_stm32_state.pcd.Init.speed = USB_OTG_SPEED_FULL;
+       usb_dc_stm32_state.pcd.Init.speed = USB_OTG_SPEED_HIGH;
 #ifdef USB_HS_PHYC
        usb_dc_stm32_state.pcd.Init.phy_itface = USB_OTG_HS_EMBEDDED_PHY;
 #else

```

**Comment by finikorg** [comment]
@ydamigos Thanks, that works, I think we need to make speed configurable.

**Comment by ydamigos** [comment]
@finikorg We could use "maximum-speed" (https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/usb/generic.txt).
```
 - maximum-speed: tells USB controllers we want to work up to a certain
			speed. Valid arguments are "super-speed", "high-speed",
			"full-speed" and "low-speed". In case this isn't passed
			via DT, USB controllers should default to their maximum
			HW capability.
```

When this PR is merged, I will work on it.

File: drivers/usb/device/usb_dc_stm32.c (hunk 1 above)
Comment (finikorg) [inline drivers/usb/device/usb_dc_stm32.c]: I think we need to change USB macro to something more readable.

**Review by finikorg — APPROVED** [review]
LGTM, with future adding USB maximum speed parameter and refactoring USB macro.

File: drivers/usb/device/usb_dc_stm32.c (hunk 1 above)
Comment (ydamigos) [reply drivers/usb/device/usb_dc_stm32.c]:
```
/*
 * USB and USB_OTG_FS are defined in STM32Cube HAL and allows to distinguish
 * between two kind of USB DC. STM32 F0, F3, and L0 series support USB device
 * controller. STM32 F4 and F7 series support USB_OTG_FS device controller.
 * STM32 F1 and L4 series support either USB or USB_OTG_FS device controller.
 *
 * WARNING: Don't mix USB defined in STM32Cube HAL and CONFIG_USB from Zephyr
 * Kconfig system.
 */
```
I know that USB macro is really confusing, that's why I have already add the above comment inside the driver.

**Pull request merged**